README.md
# json

[![CI](https://github.com/sagiegurari/c_json/workflows/CI/badge.svg?branch=master)](https://github.com/sagiegurari/c_json/actions)
[![Release](https://img.shields.io/github/v/release/sagiegurari/c_json)](https://github.com/sagiegurari/c_json/releases)
[![license](https://img.shields.io/github/license/sagiegurari/c_json)](https://github.com/sagiegurari/c_json/blob/master/LICENSE)

> JSON parser/writer for C.

* [Overview](#overview)
* [Usage](#usage)
* [Contributing](.github/CONTRIBUTING.md)
* [Release History](CHANGELOG.md)
* [License](#license)

<a name="overview"></a>
## Overview
This library provides both parsing JSON text/files and creating JSON text/files.

<a name="usage"></a>
## Usage

<!-- example source start -->
```c
#include "json.h"
#include <stdio.h>
#include <stdlib.h>


int main()
{
  // parsing the json string
  struct JsonValue *value = json_parse("{\"number\":1.6, \"null_key\" : null,\n"
                                       "\"bool_true\": true,\"bool_false\":false\n"
                                       ",\"string\": \"my string\nsecond line\" ,\n"
                                       " \"arr\": [1, 2.7, 3],  \n"
                                       "\"obj\": {\"subkey\": 88}, \"subobj\": {\"sub\":{\"subkey\": 77}}}");

  struct JsonValue *sub_value;


  // objects are converted to hashtables, so need to use the hashtable api to access/modify
  sub_value = hashtable_get(value->value->object, "string");

  // The actual value is based on the type
  printf("string value: %s\n", sub_value->value->string);

  // arrays are converted to vectors, so need to use the vector api to access/modify
  sub_value = hashtable_get(value->value->object, "arr");
  sub_value = vector_get(sub_value->value->array, 0);
  printf("array[0] value: %Lf\n", sub_value->value->number);

  // you can convert the parsed object back to string
  // The below is the same as json_stringify_with_options(value, false, 0);
  char *json_string = json_stringify(value);
  printf("JSON string:\n%s\n", json_string);
  free(json_string);

  // once done, release the parsed object
  json_release(value);

  // you can also prettify json strings which will
  // parse them and stringify them back with the
  // provided formatting options
  json_string = json_prettify("{\"key1\":1,\"key2\":\"test\",\"key3\":false,\"key4\":true,\"key5\":null,\"key6\":12.5,\"key7\":[true,false,[true, {\"sub\":1,\"a\":[true]}, {\"sub\":1,\"b\":[true,[true]]}, 1]]]", true, 2);
  printf("JSON string:\n%s\n", json_string);

  return(0);
} /* main */
```
<!-- example source end -->

## Contributing
See [contributing guide](.github/CONTRIBUTING.md)

<a name="history"></a>
## Release History

See [Changelog](CHANGELOG.md)

<a name="license"></a>
## License
Developed by Sagie Gur-Ari and licensed under the Apache 2 open source license.


CHANGELOG.md
## CHANGELOG

### v0.3.1 (2023-10-22)

* Fix unicode escape detection

### v0.3.0 (2023-09-25)

* Make numeric stringify be compatible with standard json libraries

### v0.2.0 (2023-02-06)

* Renamed release function

### v0.1.0 (2023-02-06)

* Initial release


tests/test.c
#include "test.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


void test_run(void (*fn)(void))
{
  printf("Test ... ");
  fn();
  printf("Done\n");
}


void test_fail()
{
  printf(" Error\n");
  exit(1);
}


void assert_true(bool value)
{
  assert_true_with_description(value, NULL);
}


void assert_true_with_description(bool value, char *description)
{
  if (!value)
  {
    if (description != NULL)
    {
      printf("Assert Failed, %s", description);
    }

    test_fail();
  }
}


void assert_size_equal(size_t value1, size_t value2)
{
  if (value1 != value2)
  {
#ifdef linux
    printf("Assert Failed, value: %zu not equals to value: %zu", value1, value2);
#endif
    test_fail();
  }
}


void assert_num_equal(long double value1, long double value2)
{
  if (value1 != value2)
  {
#ifdef linux
    printf("Assert Failed, value: %Lf not equals to value: %Lf", value1, value2);
#endif
    test_fail();
  }
}


void assert_string_equal(char *value1, char *value2)
{
  if (strcmp(value1, value2) != 0)
  {
    printf("Assert Failed, value: %s not equals to value: %s", value1, value2);
    test_fail();
  }
}



tests/test.h
#ifndef __TEST_H__
#define __TEST_H__

#include "json.h"
#include <stdbool.h>
#include <stddef.h>
#include <stdlib.h>

void test_run(void (*fn)());
void test_fail();
void assert_true(bool);
void assert_true_with_description(bool, char *);

void assert_size_equal(size_t, size_t);
void assert_num_equal(long double, long double);
void assert_string_equal(char *, char *);

#endif



tests/test_prettify.c
#include "test.h"


void test_impl()
{
  char *string;

  string = json_prettify("{}", true, 2);
  assert_string_equal(string, "{\n"
                      "}");
  free(string);

  string = json_prettify("{\"key1\":1,\"key2\":\"test\",\"key3\":false,\"key4\":true,\"key5\":null,\"key6\":12.5,\"key7\":[true,false,[true, {\"sub\":1,\"a\":[true]}, {\"sub\":1,\"b\":[true,[true]]}, 1]]]", true, 2);
  assert_string_equal(string, "{\n"
                      "  \"key1\": 1,\n"
                      "  \"key2\": \"test\",\n"
                      "  \"key3\": false,\n"
                      "  \"key4\": true,\n"
                      "  \"key5\": null,\n"
                      "  \"key6\": 12.5,\n"
                      "  \"key7\": [\n"
                      "    true,\n"
                      "    false,\n"
                      "    [\n"
                      "      true,\n"
                      "      {\n"
                      "        \"a\": [\n"
                      "          true\n"
                      "        ],\n"
                      "        \"sub\": 1\n"
                      "      },\n"
                      "      {\n"
                      "        \"b\": [\n"
                      "          true,\n"
                      "          [\n"
                      "            true\n"
                      "          ]\n"
                      "        ],\n"
                      "        \"sub\": 1\n"
                      "      },\n"
                      "      1\n"
                      "    ]\n"
                      "  ]\n"
                      "}");
  free(string);
} /* test_impl */


int main()
{
  test_run(test_impl);
}



src/json.c
#include "json.h"
#include "stringbuffer.h"
#include "stringfn.h"
#include <ctype.h>
#include <stdlib.h>
#include <string.h>

// private functions
static void _json_free(void *);
static void _json_release_json_array(struct Vector *);
static struct JsonValue *_json_create_null_value(void);
static struct JsonValue *_json_parse(char *, size_t /* length */, size_t * /* offset */);
static struct JsonValue *_json_parse_object(char *, size_t /* length */, size_t * /* offset */);
static struct JsonValue *_json_parse_array(char *, size_t /* length */, size_t * /* offset */);
static struct JsonValue *_json_parse_string(char *, size_t /* length */, size_t * /* offset */);
static struct JsonValue *_json_parse_number(char *, size_t /* length */, size_t * /* offset */);
static struct JsonValue *_json_parse_boolean(char *, size_t /* length */, size_t * /* offset */);
static struct JsonValue *_json_parse_null(char *, size_t /* length */, size_t * /* offset */);
static bool _json_skip_whitespaces(char *, size_t /* length */, size_t * /* offset */);
static bool _json_parse_object_key(struct HashTable *, char *, size_t /* length */, size_t * /* offset */);
static void _json_release_hashtable_key_value(char *, void *);
static bool _json_stringify(struct JsonValue *, struct StringBuffer *, bool /* multi line */, size_t /* indentation */, size_t /* current indentation */, bool /* skip indent */, struct StringBuffer * /* temporary work buffer */);
static bool _json_stringify_object(struct HashTable *, struct StringBuffer *, bool /* multi line */, size_t /* indentation */, size_t /* current indentation */, bool /* skip indent */, struct StringBuffer * /* work buffer */);
static bool _json_stringify_array(struct Vector *, struct StringBuffer *, bool /* multi line */, size_t /* indentation */, size_t /* current indentation */, bool /* skip indent */, struct StringBuffer * /* work buffer */);
static bool _json_stringify_string(char *, struct StringBuffer *);
static bool _json_stringify_number(long double, struct StringBuffer * /* output buffer */, struct StringBuffer * /* work buffer */);
static void _json_add_indentation(struct StringBuffer *, size_t /* indentation */);

struct JsonValue *json_parse(char *text)
{
  if (text == NULL)
  {
    return(NULL);
  }

  size_t length = strlen(text);
  if (!length)
  {
    return(NULL);
  }

  size_t           offset = 0;
  struct JsonValue *value = _json_parse(text, length, &offset);

  if (offset < length)
  {
    // we have some leftovers that we can't parse
    json_release(value);
    return(NULL);
  }

  return(value);
}


char *json_stringify(struct JsonValue *value)
{
  return(json_stringify_with_options(value, false /* multi line */, 0 /* indentation */));
}


char *json_stringify_with_options(struct JsonValue *value, bool multi_line, size_t indentation)
{
  if (value == NULL)
  {
    return(NULL);
  }

  struct StringBuffer *buffer      = stringbuffer_new();
  struct StringBuffer *work_buffer = stringbuffer_new_with_options(100, true);
  bool                done         = _json_stringify(value, buffer, multi_line, indentation, 0, false, work_buffer);
  stringbuffer_release(work_buffer);
  if (!done)
  {
    stringbuffer_release(buffer);
    return(NULL);
  }

  char *json_string = stringbuffer_to_string(buffer);
  stringbuffer_release(buffer);

  return(json_string);
}


char *json_prettify(char *value, bool multi_line, size_t indentation)
{
  struct JsonValue *parsed = json_parse(value);

  if (parsed == NULL)
  {
    return(NULL);
  }

  char *formatted = json_stringify_with_options(parsed, multi_line, indentation);

  json_release(parsed);

  return(formatted);
}


void json_release(struct JsonValue *value)
{
  if (value == NULL)
  {
    return;
  }

  switch (value->type)
  {
  case JSON_TYPE_OBJECT:
    hashtable_release(value->value->object);
    break;

  case JSON_TYPE_ARRAY:
    _json_release_json_array(value->value->array);
    break;

  case JSON_TYPE_STRING:
    _json_free(value->value->string);

  default:
    // no need to release
    break;
  }

  _json_free(value->value);
  _json_free(value);
}


static void _json_free(void *value)
{
  if (value == NULL)
  {
    return;
  }

  free(value);
}


static void _json_release_json_array(struct Vector *array)
{
  if (array == NULL)
  {
    return;
  }

  size_t count = vector_size(array);
  for (size_t index = 0; index < count; index++)
  {
    struct JsonValue *value = vector_get(array, index);
    json_release(value);
  }
  vector_release(array);
}

static struct JsonValue *_json_create_null_value(void)
{
  struct JsonValue *value = malloc(sizeof(struct JsonValue));

  value->type  = JSON_TYPE_NULL;
  value->value = malloc(sizeof(union JsonValueUnion));

  return(value);
}

static struct JsonValue *_json_parse(char *text, size_t length, size_t *offset)
{
  if (text == NULL || !length || *offset >= length)
  {
    return(NULL);
  }

  // skip whitespaces
  if (!_json_skip_whitespaces(text, length, offset))
  {
    return(NULL);
  }

  // detect next value type
  if (text[*offset] == '{')
  {
    return(_json_parse_object(text, length, offset));
  }

  if (text[*offset] == '[')
  {
    return(_json_parse_array(text, length, offset));
  }

  if (text[*offset] == '-' || isdigit(text[*offset]))
  {
    return(_json_parse_number(text, length, offset));
  }

  if (text[*offset] == '"')
  {
    return(_json_parse_string(text, length, offset));
  }

  if (text[*offset] == 't' || text[*offset] == 'f')
  {
    return(_json_parse_boolean(text, length, offset));
  }

  if (text[*offset] == 'n')
  {
    return(_json_parse_null(text, length, offset));
  }

  return(NULL);
} /* _json_parse */

static struct JsonValue *_json_parse_object(char *text, size_t length, size_t *offset)
{
  if (text == NULL || (*offset + 2 > length) || text[*offset] != '{')
  {
    return(NULL);
  }

  *offset = *offset + 1;

  struct HashTable *object = hashtable_new();
  struct JsonValue *value  = _json_create_null_value();
  value->type          = JSON_TYPE_OBJECT;
  value->value->object = object;

  bool found_separator      = false;
  bool can_start_next_entry = true;
  for ( ; *offset < length; *offset = *offset + 1)
  {
    // skip whitespaces
    if (!_json_skip_whitespaces(text, length, offset))
    {
      json_release(value);
      return(NULL);
    }

    size_t index = *offset;

    if (found_separator)
    {
      if (text[index] == '"')
      {
        if (!_json_parse_object_key(object, text, length, offset))
        {
          json_release(value);
          return(NULL);
        }

        // we are in next char now, but for loop will advance it one more, so go back 1
        *offset = *offset - 1;

        found_separator      = false;
        can_start_next_entry = false;
      }
      else
      {
        json_release(value);
        return(NULL);
      }
    }
    else if (!found_separator && text[index] == '}')
    {
      *offset = *offset + 1;
      break;
    }
    else if (can_start_next_entry && text[index] == '"')
    {
      if (!_json_parse_object_key(object, text, length, offset))
      {
        json_release(value);
        return(NULL);
      }

      // we are in next char now, but for loop will advance it one more, so go back 1
      *offset = *offset - 1;

      can_start_next_entry = false;
      found_separator      = false;
    }
    else if (!found_separator && text[index] == ',')
    {
      found_separator      = true;
      can_start_next_entry = true;
    }
  }

  return(value);
} /* _json_parse_object */

static struct JsonValue *_json_parse_array(char *text, size_t length, size_t *offset)
{
  if (text == NULL || (*offset + 2 > length) || text[*offset] != '[')
  {
    return(NULL);
  }

  *offset = *offset + 1;

  struct Vector    *array = vector_new_with_options(100 /* initial size */, true /* allow resize */);
  struct JsonValue *value = _json_create_null_value();
  value->type         = JSON_TYPE_ARRAY;
  value->value->array = array;

  bool found_seperator = true;
  for ( ; *offset < length; *offset = *offset + 1)
  {
    // skip whitespaces
    if (!_json_skip_whitespaces(text, length, offset))
    {
      json_release(value);
      return(NULL);
    }

    size_t index = *offset;

    if (text[index] == ']')
    {
      *offset = *offset + 1;
      break;
    }
    else if (text[index] == ',')
    {
      if (found_seperator)
      {
        json_release(value);
        return(NULL);
      }
      else
      {
        found_seperator = true;
      }
    }
    else if (found_seperator)
    {
      found_seperator = false;

      struct JsonValue *array_item = _json_parse(text, length, offset);
      if (array_item == NULL)
      {
        json_release(value);
        return(NULL);
      }

      vector_push(array, array_item);

      // we are in next char now, but for loop will advance it one more, so go back 1
      *offset = *offset - 1;
    }
    else
    {
      json_release(value);
      return(NULL);
    }
  }

  return(value);
} /* _json_parse_array */

static struct JsonValue *_json_parse_string(char *text, size_t length, size_t *offset)
{
  if (text == NULL || (*offset + 2 > length) || text[*offset] != '"')
  {
    return(NULL);
  }

  size_t              delta_offset = 0;
  bool                in_escape    = false;
  bool                found_end    = false;
  struct StringBuffer *buffer      = stringbuffer_new_with_options(100 /* initial size */, true /* allow resize */);
  for (size_t index = *offset + 1; index < length; index++)
  {
    char character = text[index];

    if (in_escape)
    {
      in_escape    = false;
      delta_offset = delta_offset + 1;

      if (character == 'b')
      {
        stringbuffer_append(buffer, '\b');
      }
      else if (character == 'f')
      {
        stringbuffer_append(buffer, '\f');
      }
      else if (character == 'n')
      {
        stringbuffer_append(buffer, '\n');
      }
      else if (character == 'r')
      {
        stringbuffer_append(buffer, '\r');
      }
      else if (character == 't')
      {
        stringbuffer_append(buffer, '\t');
      }
      else if (character == '"')
      {
        stringbuffer_append(buffer, '"');
      }
      else if (character == '\\')
      {
        stringbuffer_append(buffer, '\\');
      }
      else if (character == 'u')
      {
        stringbuffer_append_string(buffer, "\\u");
      }
      else
      {
        // invalid/unsupported escape
        stringbuffer_release(buffer);
        return(NULL);
      }
    }
    else if (character == '\\')
    {
      in_escape = true;
    }
    else if (character == '"')
    {
      found_end = true;
      break;
    }
    else
    {
      stringbuffer_append(buffer, character);
    }
  }

  if (!found_end)
  {
    stringbuffer_release(buffer);
    return(NULL);
  }

  // new offset is now previous offset + 2 (start/end ") + content length + escape characters cound
  *offset = *offset + delta_offset + stringbuffer_get_content_size(buffer) + 2;
  struct JsonValue *value = _json_create_null_value();
  value->type          = JSON_TYPE_STRING;
  value->value->string = stringbuffer_to_string(buffer);
  stringbuffer_release(buffer);

  return(value);
} /* _json_parse_string */

static struct JsonValue *_json_parse_number(char *text, size_t length, size_t *offset)
{
  if (text == NULL || (*offset + 1 > length))
  {
    return(NULL);
  }

  size_t start               = *offset;
  bool   found_decimal_point = false;
  for ( ; *offset < length; *offset = *offset + 1)
  {
    size_t index = *offset;

    char   character = text[index];
    if (isdigit(character) || (index == start && character == '-'))
    {
      // continue
    }
    else if (character == '.' && !found_decimal_point)
    {
      found_decimal_point = true;
    }
    else
    {
      break;
    }
  }

  if (start == *offset)
  {
    return(NULL);
  }

  struct JsonValue *value = _json_create_null_value();
  value->type = JSON_TYPE_NUMBER;
  char             *subtext = stringfn_substring(text, (int)start, *offset - start);
  value->value->number = strtold(subtext, NULL);
  _json_free(subtext);

  return(value);
} /* _json_parse_number */

static struct JsonValue *_json_parse_boolean(char *text, size_t length, size_t *offset)
{
  if (text == NULL)
  {
    return(NULL);
  }

  char *subtext = text + *offset;

  if ((*offset + 4 <= length) && stringfn_starts_with(subtext, "true"))
  {
    struct JsonValue *value = _json_create_null_value();
    value->type           = JSON_TYPE_BOOLEAN;
    value->value->boolean = true;
    *offset               = *offset + 4;

    return(value);
  }
  else if ((*offset + 5 <= length) && stringfn_starts_with(subtext, "false"))
  {
    struct JsonValue *value = _json_create_null_value();
    value->type           = JSON_TYPE_BOOLEAN;
    value->value->boolean = false;

    *offset = *offset + 5;

    return(value);
  }

  return(NULL);
}

static struct JsonValue *_json_parse_null(char *text, size_t length, size_t *offset)
{
  if (text == NULL || (*offset + 4 > length))
  {
    return(NULL);
  }

  char *subtext = text + *offset;

  if (stringfn_starts_with(subtext, "null"))
  {
    *offset = *offset + 4;
    return(_json_create_null_value());
  }

  return(NULL);
}


static bool _json_skip_whitespaces(char *text, size_t length, size_t *offset)
{
  if (text == NULL || *offset >= length)
  {
    return(false);
  }

  // skip whitespaces
  for ( ; *offset < length; *offset = *offset + 1)
  {
    if (!isspace(text[*offset]))
    {
      break;
    }
  }

  return(*offset < length);
}


static bool _json_parse_object_key(struct HashTable *object, char *text, size_t length, size_t *offset)
{
  if (object == NULL || text == NULL || (*offset + 4 > length) || text[*offset] != '"')
  {
    return(false);
  }

  // parse key
  struct JsonValue *json_value = _json_parse_string(text, length, offset);
  if (json_value == NULL)
  {
    return(false);
  }
  if (json_value->type != JSON_TYPE_STRING)
  {
    json_release(json_value);
    return(false);
  }

  char *key = json_value->value->string;
  _json_free(json_value->value);
  _json_free(json_value);

  // skip whitespaces
  if (!_json_skip_whitespaces(text, length, offset) || text[*offset] != ':')
  {
    _json_free(key);
    return(false);
  }
  *offset = *offset + 1;

  // parse value
  json_value = _json_parse(text, length, offset);
  if (json_value == NULL)
  {
    _json_free(key);
    return(false);
  }

  bool added = hashtable_insert(object, key, json_value, _json_release_hashtable_key_value);

  if (!added)
  {
    _json_free(key);
    json_release(json_value);
    return(false);
  }

  return(true);
} /* _json_parse_object_key */


static void _json_release_hashtable_key_value(char *key, void *value)
{
  _json_free(key);
  struct JsonValue *json_value = (struct JsonValue *)value;
  json_release(json_value);
}


static bool _json_stringify(struct JsonValue *value, struct StringBuffer *buffer, bool multi_line, size_t indentation, size_t current_indentation, bool skip_indent, struct StringBuffer *work_buffer)
{
  if (value == NULL || buffer == NULL || work_buffer == NULL)
  {
    return(false);
  }

  if (multi_line && !skip_indent)
  {
    if (!stringbuffer_is_empty(buffer))
    {
      stringbuffer_append(buffer, '\n');
    }

    if (value->type != JSON_TYPE_OBJECT && value->type != JSON_TYPE_ARRAY)
    {
      _json_add_indentation(buffer, current_indentation);
    }
  }

  switch (value->type)
  {
  case JSON_TYPE_OBJECT:
    if (!_json_stringify_object(value->value->object, buffer, multi_line, indentation, current_indentation, skip_indent, work_buffer))
    {
      return(false);
    }
    break;

  case JSON_TYPE_ARRAY:
    if (!_json_stringify_array(value->value->array, buffer, multi_line, indentation, current_indentation, skip_indent, work_buffer))
    {
      return(false);
    }
    break;

  case JSON_TYPE_STRING:
    _json_stringify_string(value->value->string, buffer);
    break;

  case JSON_TYPE_NUMBER:
    _json_stringify_number(value->value->number, buffer, work_buffer);
    break;

  case JSON_TYPE_BOOLEAN:
    stringbuffer_append_bool(buffer, value->value->boolean);
    break;

  case JSON_TYPE_NULL:
    stringbuffer_append_string(buffer, "null");
    break;
  }

  return(true);
} /* _json_stringify */


static bool _json_stringify_object(struct HashTable *object, struct StringBuffer *buffer, bool multi_line, size_t indentation, size_t current_indentation, bool skip_indent, struct StringBuffer *work_buffer)
{
  if (object == NULL || buffer == NULL || work_buffer == NULL)
  {
    return(false);
  }

  if (multi_line && !skip_indent)
  {
    _json_add_indentation(buffer, current_indentation);
  }
  stringbuffer_append(buffer, '{');

  size_t size = hashtable_size(object);
  if (size)
  {
    struct HashTableEntries entries = hashtable_entries(object);

    bool                    added            = false;
    size_t                  next_indentation = indentation + current_indentation;
    for (size_t index = 0; index < size; index++)
    {
      char             *key   = entries.keys[index];
      struct JsonValue *value = (struct JsonValue *)entries.values[index];

      if (key != NULL && value != NULL)
      {
        if (!added)
        {
          added = true;
        }
        else
        {
          stringbuffer_append(buffer, ',');
        }

        if (multi_line)
        {
          stringbuffer_append(buffer, '\n');
          _json_add_indentation(buffer, next_indentation);
        }

        if (!_json_stringify_string(key, buffer))
        {
          return(false);
        }
        stringbuffer_append(buffer, ':');
        if (multi_line)
        {
          stringbuffer_append(buffer, ' ');
        }

        if (!_json_stringify(value, buffer, multi_line, indentation, next_indentation, true, work_buffer))
        {
          return(false);
        }
      }
    }

    _json_free(entries.keys);
    _json_free(entries.values);
  }

  if (multi_line)
  {
    stringbuffer_append(buffer, '\n');
    _json_add_indentation(buffer, current_indentation);
  }
  stringbuffer_append(buffer, '}');

  return(true);
} /* _json_stringify_object */


static bool _json_stringify_array(struct Vector *array, struct StringBuffer *buffer, bool multi_line, size_t indentation, size_t current_indentation, bool skip_indent, struct StringBuffer *work_buffer)
{
  if (array == NULL || buffer == NULL || work_buffer == NULL)
  {
    return(false);
  }

  if (multi_line && !skip_indent)
  {
    _json_add_indentation(buffer, current_indentation);
  }
  stringbuffer_append(buffer, '[');

  size_t size             = vector_size(array);
  bool   added            = false;
  size_t next_indentation = indentation + current_indentation;
  for (size_t index = 0; index < size; index++)
  {
    struct JsonValue *value = vector_get(array, index);

    if (value != NULL)
    {
      if (!added)
      {
        added = true;
      }
      else
      {
        stringbuffer_append(buffer, ',');
      }

      if (!_json_stringify(value, buffer, multi_line, indentation, next_indentation, false, work_buffer))
      {
        return(false);
      }
    }
  }

  if (multi_line)
  {
    stringbuffer_append(buffer, '\n');
    _json_add_indentation(buffer, current_indentation);
  }
  stringbuffer_append(buffer, ']');

  return(true);
} /* _json_stringify_array */


static bool _json_stringify_string(char *text, struct StringBuffer *buffer)
{
  if (text == NULL || buffer == NULL)
  {
    return(false);
  }

  size_t length = strlen(text);

  stringbuffer_append(buffer, '"');
  for (size_t index = 0; index < length; index++)
  {
    char character = text[index];

    if (character == '\b')
    {
      stringbuffer_append_string(buffer, "\\b");
    }
    else if (character == '\f')
    {
      stringbuffer_append_string(buffer, "\\f");
    }
    else if (character == '\n')
    {
      stringbuffer_append_string(buffer, "\\n");
    }
    else if (character == '\r')
    {
      stringbuffer_append_string(buffer, "\\r");
    }
    else if (character == '\t')
    {
      stringbuffer_append_string(buffer, "\\t");
    }
    else if (character == '"')
    {
      stringbuffer_append_string(buffer, "\\\"");
    }
    else if (character == '\\')
    {
      stringbuffer_append_string(buffer, "\\\\");
    }
    else
    {
      stringbuffer_append(buffer, character);
    }
  }
  stringbuffer_append(buffer, '"');

  return(true);
} /* _json_stringify_string */


static bool _json_stringify_number(long double value, struct StringBuffer *buffer, struct StringBuffer *work_buffer)
{
  if (buffer == NULL || work_buffer == NULL)
  {
    return(false);
  }

  stringbuffer_clear(work_buffer);
  stringbuffer_append_long_double(work_buffer, value);
  char   *number_string   = stringbuffer_to_string(work_buffer);
  size_t length           = strlen(number_string);
  size_t length_to_remove = 0;
  bool   stop_removing    = false;
  for (size_t index = length - 1; index > 0; index--)
  {
    char character = number_string[index];
    if (character == '.')
    {
      if (length_to_remove > 0)
      {
        if (!stop_removing)
        {
          length_to_remove = length_to_remove + 1;
        }
        length = length - length_to_remove;
      }
      break;
    }
    else if (character == '0' && !stop_removing)
    {
      length_to_remove = length_to_remove + 1;
    }
    else
    {
      stop_removing = true;
    }
  }

  stringbuffer_append_string_with_options(buffer, number_string, 0, length);
  free(number_string);

  return(true);
} /* _json_stringify_number */


static void _json_add_indentation(struct StringBuffer *buffer, size_t indentation)
{
  if (buffer == NULL)
  {
    return;
  }

  for (size_t index = 0; index < indentation; index++)
  {
    stringbuffer_append(buffer, ' ');
  }
}


include/json.h
#ifndef JSON_H
#define JSON_H

#include "hashtable.h"
#include "vector.h"
#include <stdbool.h>
#include <stddef.h>

enum JsonType
{
  JSON_TYPE_OBJECT = 1,
  JSON_TYPE_ARRAY,
  JSON_TYPE_STRING,
  JSON_TYPE_NUMBER,
  JSON_TYPE_BOOLEAN,
  JSON_TYPE_NULL
};

struct JsonValue
{
  enum JsonType        type;
  union JsonValueUnion *value;
};

union JsonValueUnion
{
  struct HashTable *object;
  struct Vector    *array;
  char             *string;
  long double      number;
  bool             boolean;
};

/**
 * Parses the give string and returns the json value union.
 * In case of any error, this function will return null.
 * The json value must be fully released once done.
 */
struct JsonValue *json_parse(char * /* text */);

/**
 * Converts the provided json value to string without any special formatting.
 * In case of any error or invalid value, NULL will be returned.
 */
char *json_stringify(struct JsonValue *);

/**
 * Converts the provided json value to string with formatting based on the provided options.
 * In case of any error or invalid value, NULL will be returned.
 */
char *json_stringify_with_options(struct JsonValue *, bool /* multi line */, size_t /* indentation */);

/**
 * Parses the provided json string and stringifies it back with
 * the provided formatting options.
 */
char *json_prettify(char *, bool /* multi line */, size_t /* indentation */);

/**
 * Releases the json value and all internal memory used.
 * All internal strings will also be released, therefore no const strings
 * or freed strings must reside in the structure.
 */
void json_release(struct JsonValue *);

#endif



examples/example.c
#include "json.h"
#include <stdio.h>
#include <stdlib.h>


int main()
{
  // parsing the json string
  struct JsonValue *value = json_parse("{\"number\":1.6, \"null_key\" : null,\n"
                                       "\"bool_true\": true,\"bool_false\":false\n"
                                       ",\"string\": \"my string\nsecond line\" ,\n"
                                       " \"arr\": [1, 2.7, 3],  \n"
                                       "\"obj\": {\"subkey\": 88}, \"subobj\": {\"sub\":{\"subkey\": 77}}}");

  struct JsonValue *sub_value;


  // objects are converted to hashtables, so need to use the hashtable api to access/modify
  sub_value = hashtable_get(value->value->object, "string");

  // The actual value is based on the type
  printf("string value: %s\n", sub_value->value->string);

  // arrays are converted to vectors, so need to use the vector api to access/modify
  sub_value = hashtable_get(value->value->object, "arr");
  sub_value = vector_get(sub_value->value->array, 0);
  printf("array[0] value: %Lf\n", sub_value->value->number);

  // you can convert the parsed object back to string
  // The below is the same as json_stringify_with_options(value, false, 0);
  char *json_string = json_stringify(value);
  printf("JSON string:\n%s\n", json_string);
  free(json_string);

  // once done, release the parsed object
  json_release(value);

  // you can also prettify json strings which will
  // parse them and stringify them back with the
  // provided formatting options
  json_string = json_prettify("{\"key1\":1,\"key2\":\"test\",\"key3\":false,\"key4\":true,\"key5\":null,\"key6\":12.5,\"key7\":[true,false,[true, {\"sub\":1,\"a\":[true]}, {\"sub\":1,\"b\":[true,[true]]}, 1]]]", true, 2);
  printf("JSON string:\n%s\n", json_string);

  return(0);
} /* main */
